### DISCUSSION TITLE
How to test a non-route function that uses Depends

### DISCUSSION THREAD

**Discussion by arturoaviles** [fastapi/fastapi#6607] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

main.py
```python
from fastapi import Depends, FastAPI
from store import items

app = FastAPI()

@app.get('/items')
def get_items(i: list = Depends(items)) -> dict:
	return {"items": i}

```

store.py

```python
from fastapi import Depends

def get_items_from_db() -> list:
	db = [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]
	return db

def items(database: list = Depends(get_items_from_db)):
	return database
```

### Description

I want to test only the **items** function using pytest but I'm unable to do so.

I have tried the following (but nothing has worked):

test_store.py
```python
import fastapi
import pytest

from fastapi import Depends
from store import items


def test_store1(i: list = Depends(items)):
	assert i == [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]

def test_store2():
	i = items()
	assert i == [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]

def test_store3(monkeypatch):
	def fake_depends():
		return [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]

	monkeypatch.setattr(Depends, "__call__", fake_depends)
	i = items()
	assert i == [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]

def test_store4(monkeypatch):
	def fake_depends():
		return [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]

	monkeypatch.setattr(fastapi, "Depends", fake_depends)
	i = items()
	assert i == [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]

def test_store5(monkeypatch):
	def fake_depends():
		return [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]

	monkeypatch.setattr("fastapi.Depends", fake_depends)
	i = items()
	assert i == [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]

@pytest.fixture
def setup(monkeypatch):
	def fake_depends():
		return [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]
	monkeypatch.setattr("fastapi.Depends", fake_depends)

def test_store6(setup, i: list = Depends(items)):
	assert i == [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]

def test_store7():
	i = Depends(items)
	assert i == [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]
```

I understand that I can use the `app.dependency_overrides[get_db] = override_get_db` example  but this seems to only work when doing a test using the `client = TestClient(app)` to test the endpoints. For this case my test file is more specific.

### Operating System

Linux

### Operating System Details

Ubuntu 20.04

### FastAPI Version

0.63.0

### Python Version

3.9.5

**Comment by STeveShary** [comment, marked as answer]
The great thing about dependency injection is how simple it is to test with it!   FastAPI uses parameter injection with Python's default arguments.  When we write a test, we can pass in a dependency and python will not inject it.  It's that simple!

Let's look at your example:
```python3
def items(database: list = Depends(get_items_from_db)):
	return database
```

Instead of getting items from the database, I can just "inject" them by passing them in as a parameter.

Here's how we would do this with one of your tests that calls the function above:
```python3
def test_store1():
	expected_items =  [
			{
				"id": "1",
				"name": "chair"
			},
			{
				"id": "2",
				"name": "table"
			}
		]
	actual_items = items(expected_items)
	assert actual_items == expected_items
```

If you are interested in wiring all of your dependencies together and you want to mock out one, you can do so with dependency overrides as documented here: https://fastapi.tiangolo.com/advanced/testing-dependencies/

**Comment by arturoaviles** [comment]
Thanks @STeveShary !!! It worked.
